File: src/main/java/me/nasukhov/TukitaLearner/study/GroupQuestion.java
package me.nasukhov.TukitaLearner.study;

import jakarta.persistence.*;
import me.nasukhov.TukitaLearner.bot.io.Channel;

import java.util.ArrayList;
import java.util.List;
import java.util.UUID;

@Entity
@Table(name = "ch_questions")
public class GroupQuestion {
    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    // TODO generate uuid 7
    private UUID id;

    @Column(nullable = false)
    private String channelId;

    @ManyToOne
    private Question question;

    @OneToMany(mappedBy = "question", cascade = CascadeType.ALL, orphanRemoval = true)
    private List<Answer> answers = new ArrayList<>();

    protected GroupQuestion() {}

    // TODO could id generation encapsulation make it better?
    public GroupQuestion(Channel channel, Question question) {
        this.channelId = channel.id;
        this.question = question;
    }

    public UUID getId() {
        return id;
    }

    public String getText() {
        return question.text;
    }

    private boolean isCorrectAnswer(int numberOfAnswer) {
        String selectedAnswer = viewVariant(numberOfAnswer);
        String correctAnswer = question.answer;

        return selectedAnswer.equals(correctAnswer);
    }

    public String viewAnswer() {
        return question.answer;
    }

    public String viewVariant(int variant) {
        if (variant < 1) {
            return "";
        }

        if (variant > question.variants.size()) {
            return "";
        }

        return question.variants.get(variant-1);
    }

    public List<String> listVariants() {
        return question.variants;
    }

    public Answer addAnswer(String user, int selectedAnswer) {
        if (hasAnswerFromUser(user)) {
            throw new RuntimeException("User has already answered this question");
        }

        var answer = new Answer(
                user,
                channelId,
                this,
                isCorrectAnswer(selectedAnswer)
        );

        answers.add(answer);

        return answer;
    }

    public boolean hasAnswerFromUser(String user) {
        return answers.stream().anyMatch(answer -> answer.userId.equals(user));
    }
}

File: src/main/kotlin/me/nasukhov/tukitalearner/study/GroupQuestion.kt
package me.nasukhov.tukitalearner.study

import jakarta.persistence.*
import org.hibernate.annotations.OnDelete
import org.hibernate.annotations.OnDeleteAction
import java.util.UUID

@Entity
@Table(name = "group_questions")
class GroupQuestion {
    @Id
    @GeneratedValue(strategy = GenerationType.UUID) // TODO generate uuid 7
    lateinit var id: UUID
        private set

    @ManyToOne
    @JoinColumn(name = "channel_id", nullable = false)
    @OnDelete(action = OnDeleteAction.CASCADE)
    lateinit var group: Group

    @ManyToOne
    @JoinColumn(name = "question_id", nullable = false)
    @OnDelete(action = OnDeleteAction.CASCADE)
    private lateinit var question: Question

    val text: String
        get() = question.text

    @OneToMany(mappedBy = "question", cascade = [CascadeType.ALL], orphanRemoval = true)
    private val answers: MutableList<Answer> = ArrayList()

    // ORM necessity
    private constructor()

    constructor(group: Group, question: Question) {
        this.group = group
        this.question = question
    }

    fun viewAnswer(): String = question.answer

    fun viewVariant(variant: Int): String {
        if (variant < 1) {
            return ""
        }

        if (variant > question.variants.size) {
            return ""
        }

        return question.variants[variant - 1]
    }

    fun listVariants(): List<String> = question.variants

    fun addAnswer(
        user: String,
        selectedVariant: Int,
    ): Answer {
        check(!hasAnswerFromUser(user)) {
            "User has already answered this question"
        }

        val selectedAnswer = viewVariant(selectedVariant)
        val answer = Answer(selectedAnswer, user, this)

        answers.add(answer)

        return answer
    }

    fun hasAnswerFromUser(user: String): Boolean = answers.stream().anyMatch { answer: Answer -> answer.userId == user }
}
